src/core/algorithms/md/md_verifier/validation/validation.cpp

```diff
@@ -0,0 +1,281 @@
+            non_trivial_column_matches_info.size() + trivial_column_matches_info.size();
+    column_matches_similarity_infos_.resize(total_column_matches_number);
+
+    std::vector<model::Index> const& sorted_to_original = similarity_data.GetIndexMapping();
+    for (model::Index sorted_index : hymd::utility::IndexRange(sorted_to_original.size())) {
+        model::Index original_index = sorted_to_original[sorted_index];
+
+        column_matches_similarity_infos_[original_index] =
+                non_trivial_column_matches_info[sorted_index];
+    }
+
+    for (auto [decision_boundary, original_index] : trivial_column_matches_info) {
+        column_matches_similarity_infos_[original_index] = decision_boundary;
+    }
+}
+
+void MDValidationCalculator::ExecuteValidationFrom(model::Index lhs_classifier_index) {
+    OneOfColumnMatchInfo const& starting_column_match_info =
+            column_matches_similarity_infos_[lhs_classifier_index];
+    model::md::DecisionBoundary starting_decision_boundary =
+            column_similarity_classifiers_[lhs_classifier_index].GetDecisionBoundary();
+
+    auto on_non_trivial = [&](hymd::ColumnMatchInfo const& column_match_info) {
+        hymd::indexes::SimilarityIndex const& similarity_index =
+                column_match_info.similarity_info.similarity_index;
+
+        auto it = std::ranges::lower_bound(column_match_info.similarity_info.classifier_values,
+                                           starting_decision_boundary);
+        if (it == column_match_info.similarity_info.classifier_values.end()) {
+            // No pairs are matched by LHS, dependency holds
+            return;
+        }
+        hymd::ColumnClassifierValueId lower_bound_ccv_id =
+                std::distance(column_match_info.similarity_info.classifier_values.begin(), it);
+
+        for (hymd::ValueIdentifier left_value_id :
+             hymd::utility::IndexRange(similarity_index.size())) {
+            hymd::indexes::PliCluster const& left_pli_cluster =
+                    records_info_->GetLeftCompressor()
+                            .GetPli(column_match_info.left_column_index)
+                            .GetClusters()[left_value_id];
+            hymd::indexes::RecSet const* upper_set =
+                    similarity_index[left_value_id].GetUpperSet(lower_bound_ccv_id);
+
+            if (upper_set == nullptr) {
+                continue;
+            }
+
+            for (hymd::RecordIdentifier left_record_id : left_pli_cluster) {
+                for (hymd::RecordIdentifier right_record_id : *upper_set) {
+                    if (validation_finished_) {
+                        return;
+                    }
+                    ValidateMdConstraint(left_record_id, right_record_id);
+                }
+            }
+        }
+    };
+
+    auto on_trivial = [&](model::md::DecisionBoundary provided_decision_boundary) {
+        if (provided_decision_boundary < starting_decision_boundary) {
```

Comment:
Whoops, it's the other way around. "provided" means provided as the input parameter.
```suggestion
        if (starting_decision_boundary < provided_decision_boundary) {
```

Reply:
NOTE: swap the variable names, not the variables in this `if`